include/boost/compute/system.hpp

```diff
@@ -149,17 +160,21 @@ class system
     /// same context object being returned.
     static context default_context()
     {
-        static context default_context(default_device());
-
-        return default_context;
+        return init_default_context();
     }
 
     /// Returns the default command queue for the system.
-    static command_queue& default_queue()
+    ///
+    /// If user-provided command queue is given, the system-wide default context
+    /// and default device will be set up appropriately so that the default queue
+    /// matches the default context and device.
+    ///
+    /// The default queue is created once on the first time this function is
+    /// called. Calling this function multiple times will always result in the
+    /// same command queue object being returned.
+    static command_queue& default_queue(command_queue* user_queue = 0)
```

Comment:
Shouldn't `user_queue` be `const`? Also have we/you considered using constant references instead of pointers?

Reply:
This also applies to other functions which take pointer to queue/device/context.

Reply:
I agree it should be made `const`. As for constant reference, I don't think it's possible because the intent is to double `default_queue()` as both means to get/set the default queue. And we don't have things like null reference. 

That leaves us with two options:
1. We spin out the "set" function which takes const reference
2. Keep `default_queue()` and make the input pass-by-value, with default value set to be a null command queue

Reply:
https://github.com/boostorg/compute/blob/master/include/boost/compute/command_queue.hpp#L108 This ctor creates "null" command queue.

Reply:
Yeah that should work. Didn't figure that out myself